nexus/db-queries/src/db/datastore/snapshot_replacement.rs

```diff
@@ -0,0 +1,1428 @@
+
+                diesel::delete(
+                    volume_repair_dsl::volume_repair.filter(
+                        volume_repair_dsl::repair_id
+                            .eq(snapshot_replacement_step_id),
+                    ),
+                )
+                .execute_async(&conn)
+                .await?;
+
+                use db::schema::snapshot_replacement_step::dsl;
+                let result = diesel::update(dsl::snapshot_replacement_step)
+                    .filter(dsl::id.eq(snapshot_replacement_step_id))
+                    .filter(dsl::operating_saga_id.eq(operating_saga_id))
+                    .filter(dsl::old_snapshot_volume_id.is_null())
+                    .filter(
+                        dsl::replacement_state
+                            .eq(SnapshotReplacementStepState::Running),
+                    )
+                    .set((
+                        dsl::replacement_state
+                            .eq(SnapshotReplacementStepState::Complete),
+                        dsl::operating_saga_id.eq(Option::<Uuid>::None),
+                        dsl::old_snapshot_volume_id.eq(old_snapshot_volume_id),
+                    ))
+                    .check_if_exists::<SnapshotReplacementStep>(
+                        snapshot_replacement_step_id,
+                    )
+                    .execute_and_check(
+                        &*self.pool_connection_authorized(opctx).await?,
+                    )
+                    .await?;
+
+                match result.status {
+                    UpdateStatus::Updated => Ok(()),
+                    UpdateStatus::NotUpdatedButExists => {
+                        let record = result.found;
+
+                        if record.operating_saga_id == None
+                            && record.replacement_state
+                                == SnapshotReplacementStepState::Complete
+                        {
+                            Ok(())
+                        } else {
+                            Err(TxnError::CustomError(Error::conflict(
+                                format!(
+                                    "snapshot replacement step {} set to {:?} \
+                                (operating saga id {:?})",
+                                    snapshot_replacement_step_id,
+                                    record.replacement_state,
+                                    record.operating_saga_id,
+                                ),
+                            )))
+                        }
+                    }
+                }
+            })
+            .await
+            .map_err(|e| match e {
+                TxnError::CustomError(error) => error,
+
+                TxnError::Database(error) => {
+                    public_error_from_diesel(error, ErrorHandler::Server)
+                }
+            })
+    }
+
+    /// Count all in-progress snapshot replacement steps for a particular
+    /// snapshot replacement id.
```

Comment:
How many snapshot replacement steps do we expect to find here?
Is this like a count of progress?  The more the steps, the further along we are?

Reply:
It's one replacement step per volume that copied the snapshot volume, so for heavily copied snapshots this could be a lot. If this function returns non-zero then there's still more work to do (the coming omdb PR reports this number so we can see what's left)